### PR TITLE
Add configuration options to apply labels to log events

### LINKED ISSUE #2857: Support agent tags/labels for forwarded logs
### Description
Customers have asked for the ability to configure custom tags/labels that get included with forwarded logs existing agent behavior comes close to what's been requested.

### Acceptance Criteria
When custom tags/labels are configured for the agent, those get attached to forwarded log as attributes

### Design Consideration/Limitations
N/A

### Dependencies
N/A

### Additional context
Review epic description on [NR-315883](https://new-relic.atlassian.net/browse/NR-315883)

Similar Jiras on this topic - [NR-303269](https://new-relic.atlassian.net/browse/NR-303269), [NR-310852](https://new-relic.atlassian.net/browse/NR-310852) & [NR-312135](https://new-relic.atlassian.net/browse/NR-312135)

[NR-303269]: https://new-relic.atlassian.net/browse/NR-303269?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-310852]: https://new-relic.atlassian.net/browse/NR-310852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NR-312135]: https://new-relic.atlassian.net/browse/NR-312135?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[NR-315883]: https://new-relic.atlassian.net/browse/NR-315883?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### PR DESCRIPTION
**Pull request by kaylareopelle**
Add your New Relic labels/tags to your log events

Now, logs forwarded by the Ruby agent can have your New Relic labels (referred to as tags in the UI), as set using the `:labels` config option in `newrelic.yml` or the `NEW_RELIC_LABELS` environment variable, applied as attributes to your log events. This can help you cross-reference information you're used to seeing on your entities and transactions on your log events as well.

All labels attached to log events as attributes will be prefixed with `tags.`

This feature is off by default. To enable it, you will need to update your configuration:
newrelic.yml:
`application_logging.forwarding.labels.enabled: true` 

environment variable:
`NEW_RELIC_APPLICATION_LOGGING_FORWARDING_LABELS_ENABLED=true`

In addition, we have a configuration option that allows you to specify labels you do not want to attach to your log events. `application_logging.forwarding.labels.exclude` is a case-insensitive list of label names you'd like excluded from the log events.

For example, the following configuration:
```yaml
# newrelic.yml

labels: 'Server:One;Data Center:Primary;environment:staging;cell:grapes'
application_logging.forwarding.enabled: true
application_logging.forwarding.labels.enabled: true
application_logging.forwarding.labels.exclude:
  - server
```

would add these attributes to your log event: 
```
tags.Data Center = 'Primary'
tags.environment = 'staging'
tags.cell = 'grapes'
```

Resolves #2857

### CODE REVIEW

File: lib/new_relic/agent/log_event_aggregator.rb [hunk 1]
```diff
@@ -201,8 +208,9 @@ def self.payload_to_melt_format(data)
         # To save on unnecessary data transmission, trim the entity.type
         # sent by classic logs-in-context
         common_attributes.delete(ENTITY_TYPE_KEY)
-
-        common_attributes.merge!(NewRelic::Agent.agent.log_event_aggregator.attributes.custom_attributes)
+        aggregator = NewRelic::Agent.agent.log_event_aggregator
+        common_attributes.merge!(aggregator.attributes.custom_attributes)
+        common_attributes.merge!(aggregator.labels)
```
Comment (kaylareopelle) [inline lib/new_relic/agent/log_event_aggregator.rb:213]: This leaves me wondering if we should leverage the LogEventAttributes class for the labels work too, rather than leaving all the labels work inside the aggregator. However, this may make labels more difficult to remove when the aggregator is reset. Thoughts?

File: test/new_relic/agent/log_event_aggregator_test.rb [hunk 2]
```diff
@@ -324,6 +328,35 @@ def test_records_metrics_on_harvest
       end
     end
 
+    def test_seen_and_sent_zeroed_out_on_reset!
```
Comment (kaylareopelle) [inline test/new_relic/agent/log_event_aggregator_test.rb:331]: Since I updated `reset!`, I thought it could use more testing for its other functionality.

File: test/new_relic/agent/log_event_aggregator_test.rb [hunk 3]
```diff
@@ -343,14 +376,24 @@ def test_high_security_mode
...
+      # set the overall enabled key to false
+      # put all other configs as true
+      with_config(
+        LogEventAggregator::OVERALL_ENABLED_KEY => false,
+        :'instrumentation.logger' => 'auto',
+        LogEventAggregator::METRICS_ENABLED_KEY => true,
+        LogEventAggregator::FORWARDING_ENABLED_KEY => true,
+        LogEventAggregator::DECORATING_ENABLED_KEY => true,
+        LogEventAggregator::LABELS_ENABLED_KEY => true
+      ) do
```
Comment (kaylareopelle) [inline test/new_relic/agent/log_event_aggregator_test.rb:396]: This is unrelated to the labels change. Happy to put it in a separate PR. Essentially, I wanted to make sure that the overall enabled shutoff really flipped the values for all related configs. Since labels are disabled by default, the disabled assertion below felt like a potential false positive.

File: lib/new_relic/agent/log_event_aggregator.rb [hunk 4]
```diff
@@ -222,6 +230,7 @@ def reset!
         @counter_lock.synchronize do
           @seen = 0
           @seen_by_severity.clear
+          remove_instance_variable(:@labels) if instance_variable_defined?(:@labels)
```
Comment (kaylareopelle) [inline lib/new_relic/agent/log_event_aggregator.rb:233]: I don't know if I like this or even want it. It came up because the labels leaked between a few tests that weren't specifically about labels. Perhaps clearing things out in those tests is more appropriate.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,11 +2,15 @@
...
 
 - **Feature: Instrumentation for aws-sdk-lambda**
 
-  If the aws-sdk-lambda gem is present and used to invoke remote AWS Lambda functions, timing and error details for the invocations will be reported to New Relic. [PR#2926](https://github.com/newrelic/newrelic-ruby-agent/pull/2926)
+  If the aws-sdk-lambda gem is present and used to invoke remote AWS Lambda functions, timing and error details for the invocations will be reported to New Relic. [PR#2926](https://github.com/newrelic/newrelic-ruby-agent/pull/2926).
+
+- **Feature: Add new configuration options to attach custom tags (labels) to logs**
+
+  The Ruby agent now allows you to opt-in to adding your custom tags (labels) to agent-forwarded logs. With custom tags on logs, platform engineers can easily filter, search, and correlate log data for faster and more efficient troubleshooting, improved performance, and optimized resource utilization. To learn more about this feature see the documentation: </docs/logs/logs-context/Custom-tags-agent-forwarder-logs>. [PR#2925](https://github.com/newrelic/newrelic-ruby-agent/pull/2925)
```
Comment (kaylareopelle) [inline CHANGELOG.md:13]: This needs to be updated with the actual link before we merge.

File: lib/new_relic/agent/log_event_aggregator.rb (hunk 4 above)
Comment (hannahramadan) [reply lib/new_relic/agent/log_event_aggregator.rb:233]: Oooo so this edit is to help leaky tests? It does seem to me like that the edit would make more sense to live with the tests if so.

Comment (tannalynn) [reply lib/new_relic/agent/log_event_aggregator.rb:233]: Is this something that could also be a problem IRL, and just pops up in the tests, or is this something that will only ever be a test problem?

File: test/new_relic/agent/log_event_aggregator_test.rb [hunk 6]
```diff
@@ -324,6 +328,35 @@ def test_records_metrics_on_harvest
...
+        },
+          :ignore_filter => %r{^Supportability/API/})
+
+        @aggregator.record('Are you counting this?', 'DEBUG')
+        @aggregator.reset! # set seen/sent to zero, throwing out the latest #record
+        @aggregator.harvest! # nothing new should be available to count, so use same numbers as before
+
+        assert_metrics_recorded_exclusive({
+          'Logging/lines' => {:call_count => 3},
+          'Logging/lines/DEBUG' => {:call_count => 3}, # ?
```
Comment (tannalynn) [inline test/new_relic/agent/log_event_aggregator_test.rb:351]: why the ? comment?

File: CHANGELOG.md (hunk 5 above)
Comment (kaylareopelle) [inline CHANGELOG.md:13]:
```suggestion
  The Ruby agent now allows you to opt-in to adding your custom tags (labels) to agent-forwarded logs. With custom tags on logs, platform engineers can easily filter, search, and correlate log data for faster and more efficient troubleshooting, improved performance, and optimized resource utilization. [PR#2925](https://github.com/newrelic/newrelic-ruby-agent/pull/2925)
```

Comment (kaylareopelle) [reply CHANGELOG.md:13]: We're going to leave out the URL and may retroactively add it to the entry.

File: lib/new_relic/agent/log_event_aggregator.rb (hunk 4 above)
Comment (kaylareopelle) [reply lib/new_relic/agent/log_event_aggregator.rb]: We talked about this synchronously and determined this isn't needed in the agent, just the tests.

File: test/new_relic/agent/log_event_aggregator_test.rb (hunk 6 above)
Comment (kaylareopelle) [reply test/new_relic/agent/log_event_aggregator_test.rb:351]: ooh, forgot about this one, good catch. I felt skeptical that those were the correct numbers, but since we're resetting before harvesting, they are the correct numbers.

**Review by tannalynn — APPROVED** [review]

**Review by hannahramadan — APPROVED** [review]
⚡️
